server/ServerFSM.cpp

```diff
@@ -1785,3 +1815,125 @@ sc::result ProcessingTurn::react(const CheckTurnEndConditions& c) {
     if (TRACE_EXECUTION) DebugLogger() << "(ServerFSM) ProcessingTurn.CheckTurnEndConditions";
     return discard_event();
 }
+
+
+////////////////////////////////////////////////////////////
+// ShuttingDownServer
+////////////////////////////////////////////////////////////
+const auto SHUTDOWN_POLLING_TIME = std::chrono::milliseconds(5000);
+
+ShuttingDownServer::ShuttingDownServer(my_context c) :
+    my_base(c),
+    m_player_id_ack_expected(),
+    m_timeout(Server().m_io_service, Clock::now() + SHUTDOWN_POLLING_TIME)
+{
+    if (TRACE_EXECUTION) DebugLogger() << "(ServerFSM) ShuttingDownServer";
+
+    ServerApp& server = Server();
+
+    if (server.m_ai_client_processes.empty() && server.m_networking.empty())
+        throw ServerApp::NormalExitException();
+
+    DebugLogger() << "ShuttingDownServer informing AIs game is ending";
+
+    // Inform all players that the game is ending.  Only check the AIs for acknowledgement, because
+    // they are the server's child processes.
+    for (PlayerConnectionPtr player : server.m_networking) {
```

Comment:
can a player disconnecting during this loop change the contents of m_networking?